mxfusion/inference/batch_loop.py

```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
 
         if logger:
             logger.close()
+
+
+class BatchInferenceLoopScipy(GradLoop):
+    """
+    The class for the main loop for batch gradient-based optimization.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._initialized = None
+
+    def run(self, infr_executor, data, param_dict, ctx, optimizer='adam', learning_rate=None, max_iter=1000,
+            verbose=False, logger=None):
+        """
+        :param learning_rate: Not used
+        :param infr_executor: The MXNet function that computes the training objective.
+        :type infr_executor: MXNet Gluon Block
+        :param data: a list of observed variables
+        :type data: [mxnet.ndarray]
+        :param param_dict: The MXNet ParameterDict for Gradient-based optimization
+        :type param_dict: mxnet.gluon.ParameterDict
+        :param ctx: MXNet context
+        :type ctx: mxnet.cpu or mxnet.gpu
+        :param optimizer: the choice of optimizer (default: 'adam')
+        :type optimizer: str
+        :param max_iter: the maximum number of iterations of gradient optimization
+        :type max_iter: int
+        :param verbose: whether to print per-iteration messages.
+        :type verbose: boolean
+        :param logger: The logger to send logs to
+        :type logger: :class:`inference.Logger`
+        """
+
+        if self._initialized is None:
+            self._initialize(param_dict)
+
+        x0 = self._extract_array_from_param_dict(param_dict)
+        func = lambda x: self._f_df(x, infr_executor, data, ctx, param_dict)
+
+        x_opt, f_opt, _ = fmin_l_bfgs_b(func, x0=x0, maxiter=max_iter)
+        print('Final loss:', f_opt)
+
+    def _f_df(self, x, infr_executor, data, ctx, param_dict):
+        """
+        Objective function for optimization. Returns function value and gradient
+        """
+        self._set_param_values_from_array(x, param_dict)
+        with mx.autograd.record():
+            loss, loss_for_gradient = infr_executor(mx.nd.zeros(1, ctx=ctx), *data)
+            loss_for_gradient.backward()
+        return loss.asnumpy().astype('float64'), self._extract_grad(param_dict).astype('float64')
+
+    def _extract_grad(self, param_dict):
+        """
+        Extracts gradient information from parameter dictionary
+        """
+        x = mx.nd.zeros(self._i_end[-1])
+        for i, (key, val) in enumerate(param_dict.items()):
+            x[self._i_start[i]:self._i_end[i]] = val.data().grad.reshape((-1,))
+        return x.asnumpy()
+
+    def _initialize(self, param_dict):
+        """
+        Stores stuff like shapes of all the parameters so we can flatten them all to one array for scipy and reshape
+        them back to the correct shape
+        """
```

Comment:
Initialize and cache the shape and offset of individual parameters seem unnecessary and prone to error, e.g., a user may fix or unfix an array between runs. I would suggest to compute this on-the-fly when the run function is called.

Reply:
Good point, I'll compute these things every time the `run` method is called.